keps/sig-storage/3751-volume-attributes-class/README.md

```diff
@@ -208,30 +231,44 @@ The CSI create request will be extended to add mutable parameters. A new Control
 // ControllerServer is the server API for Controller service.
 type ControllerServer interface {
     ...
-    rpc ControllerModifyVolume (ModifyVolumeRequest)
-        returns (ModifyVolumeResponse) {
+    rpc ControllerModifyVolume (ControllerModifyVolumeRequest)
+        returns (ControllerModifyVolumeResponse) {
             option (alpha_method) = true;
         }
     ...
 }
 
-message ModifyVolumeRequest {
+message ControllerModifyVolumeRequest {
   // Contains identity information for the existing volume.
   // This field is REQUIRED.
-  string volume_id = 1
-  // This field is OPTIONAL.This allows the CO to specify the
-  // mutable parameters to apply. 
-  map<string, string> mutable_parameters = 2;
+  string volume_id = 1;
+
+  // Secrets required by plugin to complete modify volume request.
+  // This field is OPTIONAL. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 2 [(csi_secret) = true];
+
+  // Plugin specific volume attributes to mutate, passed in as 
+  // opaque key-value pairs. 
+  // This field is REQUIRED. The Plugin is responsible for
+  // parsing and validating these parameters. COs will treat these
+  // as opaque. The CO SHOULD specify the intended values of all mutable
+  // parameters it intends to modify. SPs MUST NOT modify volumes based
+  // on the absence of keys, only keys that are specified should result
+  // in modifications to the volume.
+  map<string, string> mutable_parameters = 3;
```

Comment:
Is the parameters field of the VAC optional? If the mutable_parameters is required, the field in the VAC definition may be required as well and at least contains one key.

from chatgpt:

<img width="731" alt="企业微信截图_31b09e72-e351-4243-b8c3-dcdd7e993fb8" src="https://github.com/kubernetes/enhancements/assets/28390961/9a8bf977-f2b9-4363-97a4-52de3de5ecea">

Reply:
I’m curious why the required and optional keyword is not used in fields in the csi proto file.

Reply:
We can add it, but that is the CSI spec update. The CSI Spec release has been cut already. We can revisit at next CSI meeting.

Reply:
Actually in the CSI spec, there are some extra paragraph I just did not copy pasted here because it is stand alone:

```
// Plugin specific creation-time parameters passed in as opaque 
  // key-value pairs. These mutable_parameteres MAY also be
  // changed during the lifetime of the volume via a subsequent
  // `ControllerModifyVolume` RPC. This field is OPTIONAL.
  // The Plugin is responsible for parsing and validating these
  // parameters. COs will treat these as opaque.
```

Please refer to the CSI spec for accuracy.

Reply:
And also in CSI spec:

```
message ControllerModifyVolumeRequest {
  option (alpha_message) = true;

  // Contains identity information for the existing volume.
  // This field is REQUIRED.
  string volume_id = 1;

  // Secrets required by plugin to complete modify volume request.
  // This field is OPTIONAL. Refer to the `Secrets Requirements`
  // section on how to use this field.
  map<string, string> secrets = 2 [(csi_secret) = true];

  // Plugin specific volume attributes to mutate, passed in as 
  // opaque key-value pairs. 
  // This field is REQUIRED. The Plugin is responsible for
  // parsing and validating these parameters. COs will treat these
  // as opaque. The CO SHOULD specify the intended values of all mutable
  // parameters it intends to modify. SPs MUST NOT modify volumes based
  // on the absence of keys, only keys that are specified should result
  // in modifications to the volume.
  map<string, string> mutable_parameters = 3;
}

```

Reply:
So in CreateVolume it is optional, but in ModifyVolume it is required. I think that makes sense.

From a k8s POV, if a user doesn't specify any vacName in the PVC, then that results in empty mutable_parameters.

A VAC having empty parameters doesn't make sense. If there is nothing to change then why would someone specify a vacName in their PVC?

Reply:
I am going to remove this session in the doc and please refer to the CSI spec.md